### PR TITLE
Prototype/macro effect extraction

### PR DESCRIPTION
**Pull request by lmlynik**
POC: Effect abstraction from workflows4s-core

  ⚠️ This is a proof-of-concept branch. It may not compile fully and is intended for exploration/discussion.

  Overview

  This branch explores abstracting the cats-effect dependency from workflows4s-core to make the library effect-polymorphic. The goal is to allow users to use
  workflows4s with different effect systems (cats-effect IO, ZIO, etc.) without core depending on any specific implementation.

  Approach

  - Effect type class (Effect[F[_]]) in core providing minimal operations needed for workflows: monadic ops, error handling, time operations, and delay
  - HasEffect macro using transparent inline to extract the effect type F[_] from WorkflowContext at compile time
  - Type erasure pattern - effects stored as Any internally, cast back at call sites using the extracted type
  - Separate modules: workflows4s-core (abstract), workflows4s-cats-effect (IO implementation)

  Key changes

  - WorkflowContext now has abstract type member F[_] for the effect type
  - WorkflowInstanceBase supports different effect types for instance vs engine via liftEngineEffect
  - WorkflowInstanceEngine, KnockerUpper.Agent, WorkflowRegistry.Agent are parameterized by effect type
  - Custom Effect instances can be defined (e.g., Effect[Kleisli[ConnectionIO, LiftIO, *]] for doobie)

  Known limitations / TODO

  - Some Cats abstractions like Defer and Resource need further work to be properly abstracted
  - Tests that depended on WorkflowRuntimeTest.Suite have been simplified with TODOs for restoration once IO-specific test infrastructure is available
  - The macro approach may have edge cases not yet discovered

  Related

  Addresses exploration for effect-polymorphism discussed in the project.

### CODE REVIEW

File: workflows4s-cats-effect/src/main/scala/workflows4s/runtime/instanceengine/WorkflowInstanceEngineBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import cats.effect.IO
+import workflows4s.catseffect.CatsEffect.given
+import workflows4s.effect.Effect
+import workflows4s.runtime.registry.WorkflowRegistry
+import workflows4s.runtime.wakeup.KnockerUpper
+
+import java.time.Clock
+
+object WorkflowInstanceEngineBuilder {
```
Comment (Krever) [inline workflows4s-cats-effect/src/main/scala/workflows4s/runtime/instanceengine/WorkflowInstanceEngineBuilder.scala:11]: This probably belongs to core, with `withCatsTime` as an extension. Or we can think of alternative API, but the point stands

File: workflows4s-cats-effect/src/main/scala/workflows4s/runtime/instanceengine/WorkflowInstanceEngineBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+      }
+
+    }
+
+  }
+
+  // Specialized builder for IO to avoid type parameter when using withCatsTime
+  class Step1IO(val get: WorkflowInstanceEngine[IO]) {
```
Comment (Krever) [inline workflows4s-cats-effect/src/main/scala/workflows4s/runtime/instanceengine/WorkflowInstanceEngineBuilder.scala:47]: Not sure if its needed

File: workflows4s-cats-effect/src/main/scala/workflows4s/runtime/InMemorySyncRuntime.scala [hunk 3]
```diff
@@ -31,8 +29,8 @@ object InMemorySyncRuntime {
   def create[Ctx <: WorkflowContext](
       workflow: Initial[Ctx],
       initialState: WCState[Ctx],
-      engine: WorkflowInstanceEngine,
+      engine: WorkflowInstanceEngine[Id],
       templateId: String = s"in-memory-sync-runtime-${java.util.UUID.randomUUID().toString.take(8)}",
   ): InMemorySyncRuntime[Ctx] =
-    new InMemorySyncRuntime[Ctx](workflow, initialState, engine, templateId)(using IORuntime.global)
+    new InMemorySyncRuntime[Ctx](workflow, initialState, engine, templateId)
 }
```
Comment (Krever) [inline workflows4s-cats-effect/src/main/scala/workflows4s/runtime/InMemorySyncRuntime.scala:36]: nice

File: workflows4s-cats-effect/src/main/scala/workflows4s/runtime/InMemorySyncWorkflowInstance.scala [hunk 4]
```diff
@@ -14,21 +13,21 @@ import scala.util.chaining.scalaUtilChainingOps
...
 
-  override protected def liftIO: cats.effect.LiftIO[Id]                   = new LiftIO[Id] {
-    override def liftIO[A](ioa: IO[A]): Id[A] = ioa.unsafeRunSync()
-  }
-  override protected def fMonad: Monad[Id]                                = cats.Invariant.catsInstancesForId
+  override protected def liftEngineEffect[A](fa: Id[A]): Id[A] = fa
+
   override protected def getWorkflow: workflows4s.wio.ActiveWorkflow[Ctx] = wf
 
   private val lock = new Object
```
Comment (Krever) [inline workflows4s-cats-effect/src/main/scala/workflows4s/runtime/InMemorySyncWorkflowInstance.scala:33]: I think this class would have to look somewhat differently - one of the reason this runtime exists is to take a proper workflow (wirtten with `IO` or whatever) and simplify it into `Id`. 

I think engine effect will have to be a parameter and the functionK to lift it as well.

File: workflows4s-core/src/main/scala/workflows4s/effect/concurrent/AtomicCell.scala [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  trait Make[F[_]] {
+
+    /** Creates a new AtomicCell initialized with the given value. */
+    def of[A](initial: A): F[AtomicCell[F, A]]
+  }
+
+  object Make {
+    def apply[F[_]](using m: Make[F]): Make[F] = m
+  }
+}
```
Comment (Krever) [inline workflows4s-core/src/main/scala/workflows4s/effect/concurrent/AtomicCell.scala:43]: Reminder for myself: check where its used/needed

File: workflows4s-core/src/main/scala/workflows4s/effect/concurrent/Concurrent.scala [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+  /** Computation completed successfully with a value wrapped in F. */
+  case Succeeded[G[_], B](fa: G[B]) extends Outcome[G, B]
+
+  /** Computation failed with an exception. */
+  case Errored[G[_], B](e: Throwable) extends Outcome[G, B]
+
+  /** Computation was cancelled. */
+  case Canceled[G[_], B]() extends Outcome[G, B]
+}
```
Comment (Krever) [inline workflows4s-core/src/main/scala/workflows4s/effect/concurrent/Concurrent.scala:53]: Do we really need all of those?

File: workflows4s-core/src/main/scala/workflows4s/runtime/instanceengine/BasicEngine.scala [hunk 7]
```diff
@@ -1,39 +1,38 @@
...
-import cats.implicits.catsSyntaxApplicativeId
+import workflows4s.effect.Effect
 import workflows4s.wio.*
 import workflows4s.wio.internal.{SignalResult, WakeupResult}
 import workflows4s.wio.model.WIOExecutionProgress
 
 import java.time.Instant
 
-trait BasicEngine extends WorkflowInstanceEngine {
+abstract class BasicEngine[F[_]](using E: Effect[F]) extends WorkflowInstanceEngine[F] {
```
Comment (Krever) [inline workflows4s-core/src/main/scala/workflows4s/runtime/instanceengine/BasicEngine.scala:10]: traits can have parameters in Scala 3. Do we still need abstract class?

File: workflows4s-core/src/main/scala/workflows4s/runtime/instanceengine/GreedyWorkflowInstanceEngine.scala [hunk 8]
```diff
@@ -1,13 +1,14 @@
...
 import workflows4s.wio.{ActiveWorkflow, WorkflowContext}
 
-class GreedyWorkflowInstanceEngine(protected val delegate: WorkflowInstanceEngine) extends DelegatingWorkflowInstanceEngine {
+class GreedyWorkflowInstanceEngine[F[_]](protected val delegate: WorkflowInstanceEngine[F])(using val E: Effect[F])
+    extends DelegatingWorkflowInstanceEngine[F] {
 
-  override def onStateChange[Ctx <: WorkflowContext](oldState: ActiveWorkflow[Ctx], newState: ActiveWorkflow[Ctx]): IO[Set[PostExecCommand]] = {
-    super.onStateChange(oldState, newState).map(_ + PostExecCommand.WakeUp)
+  override def onStateChange[Ctx <: WorkflowContext](oldState: ActiveWorkflow[Ctx], newState: ActiveWorkflow[Ctx]): F[Set[PostExecCommand]] = {
+    E.map(super.onStateChange(oldState, newState))(_ + PostExecCommand.WakeUp)
```
Comment (Krever) [inline workflows4s-core/src/main/scala/workflows4s/runtime/instanceengine/GreedyWorkflowInstanceEngine.scala:11]: extension would be nice

File: workflows4s-core/src/main/scala/workflows4s/runtime/WorkflowInstanceBase.scala [hunk 9]
```diff
@@ -1,52 +1,61 @@
...
-      } yield result
+      E.flatMap(liftEngineEffect(engine.handleSignal(state, signalDef, req))) { result =>
+        if !result.hasEffect then E.pure(Left(WorkflowInstance.UnexpectedSignal(signalDef)))
+        else {
+          val processed = result.asInstanceOf[SignalResult.Processed[EngineF, WCEvent[Ctx], Resp]]
+          E.flatMap(liftEngineEffect(processed.resultIO)) { eventAndResp =>
+            E.flatMap(persistEvent(eventAndResp.event)) { _ =>
+              E.flatMap(liftEngineEffect(engine.handleEvent(state, eventAndResp.event))) { newStateOpt =>
+                E.flatMap(newStateOpt.fold(E.unit)(updateState)) { _ =>
+                  E.map(handleStateChange(state, newStateOpt))(_ => Right(eventAndResp.response))
```
Comment (Krever) [inline workflows4s-core/src/main/scala/workflows4s/runtime/WorkflowInstanceBase.scala:52]: We definitely need flatMap syntax for for-comprehensions :D

File: workflows4s-core/src/main/scala/workflows4s/wio/builders/HandleSignalBuilder.scala [hunk 10]
```diff
@@ -20,12 +19,20 @@ object HandleSignalBuilder {
...
       class Step1[Input] {
 
-        def withSideEffects[Evt <: WCEvent[Ctx]](f: (Input, Req) => IO[Evt])(using evtCt: ClassTag[Evt]): Step2[Evt] = Step2(f, evtCt)
-
-        def purely[Evt <: WCEvent[Ctx]](f: (Input, Req) => Evt)(using evtCt: ClassTag[Evt]): Step2[Evt] =
-          Step2((x, y) => IO.pure(f(x, y)), evtCt)
+        /** Handle signal with side effects. The F parameter should match the workflow context's effect type. The function should return F[Evt] where
+          * F is the effect type defined in the WorkflowContext.
+          */
+        def withSideEffects[F[_], Evt <: WCEvent[Ctx]](f: (Input, Req) => F[Evt])(using evtCt: ClassTag[Evt]): Step2[Evt] =
```
Comment (Krever) [inline workflows4s-core/src/main/scala/workflows4s/wio/builders/HandleSignalBuilder.scala:25]: You have Ctx in the first step, this should allow us to extract F from there, no take arbitrary one

File: workflows4s-core/src/main/scala/workflows4s/wio/builders/RunIOBuilder.scala [hunk 11]
```diff
@@ -15,11 +14,20 @@ object RunIOBuilder {
...
     class RunIOBuilderStep1[Input] {
-      def apply[Evt <: WCEvent[Ctx]: ClassTag](f: Input => IO[Evt]): Step2[Input, Evt] = {
-        new Step2[Input, Evt](f)
+
+      /** Build a RunIO step with effect type from WorkflowContext.
+        *
+        * Uses transparent inline + HasEffect to ensure type safety: the function must return `F[Evt]` where F is the context's effect type. The
+        * effect function is stored type-erased and cast back at runtime by the evaluator.
+        */
+      transparent inline def apply[Evt <: WCEvent[Ctx]: ClassTag](using
```
Comment (Krever) [inline workflows4s-core/src/main/scala/workflows4s/wio/builders/RunIOBuilder.scala:23]: Lets hope those transparent inlines wont be necessary in the end
